### ISSUE TITLE
Add support for advanced context in request/response interceptors

### ISSUE THREAD

**Issue by vojtechvit** [RSuter/NSwag#1061]
Current Behavior
----------------------
Currently the request interception methods generated are:
```c#
partial void PrepareRequest(System.Net.Http.HttpClient client, System.Net.Http.HttpRequestMessage request, string url);
partial void PrepareRequest(System.Net.Http.HttpClient client, System.Net.Http.HttpRequestMessage request, System.Text.StringBuilder urlBuilder);
partial void ProcessResponse(System.Net.Http.HttpClient client, System.Net.Http.HttpResponseMessage response);
```

The Problem
----------------------------
The existing interception mechanism gives very little context about the API that is being called (basically just the URL), which blocks implementation of many useful or even critical interceptions.

For example - in one our Swagger-based API we store the OAuth scopes required on each endpoint in [OpenAPI Flow objects](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#oauth-flow-object). Hence, we need to extend the generated proxy to add the correct token to the correct endpoint. With the current approach, this is not possible.

Suggested solution
-------------------
The generated interception methods should be passed a "context" object that should contain all information about the endpoint called that was extracted from the OpenAPI specification. The signature of those methods could then look something like this:

```c#
partial void PrepareRequest(System.Net.Http.HttpClient client, System.Net.Http.HttpRequestMessage request, OpenApiRequestContext context);
partial void ProcessResponse(System.Net.Http.HttpClient client, System.Net.Http.HttpResponseMessage response, OpenApiRequestContext context);
```
where
```c#
public class OpenApiRequestContext
{
    public OpenApiContext Api { get; }
    public OpenApiEndPointContext EndPoint { get; }
}
```

I also believe that the interception methods should use Async pattern because it might be necessary to perform an I/O operation to get the access token. Hence, this would be even better:

```c#
partial Task PrepareRequestAsync(System.Net.Http.HttpClient client, System.Net.Http.HttpRequestMessage request, OpenApiRequestContext context);
partial Task ProcessResponseAsync(System.Net.Http.HttpClient client, System.Net.Http.HttpResponseMessage response, OpenApiRequestContext context);
```

**Comment by RSuter** [comment]
Ref: https://github.com/RSuter/NSwag/issues/321

**Comment by RSuter** [comment]
Generally a good idea. I think we should add this with a new setting (e.g. GenerateExtensionPoints). An async method cannot be partial because a partial method cannot return something (also not a Task). I think we should just call PrepareRequestAsync or ProcessResponseAsync and assume it is implemented on the specified base class... What do you expect for OpenApiContext? This looks like we have to generate lots of code... I dont like generating too much code :-)

**Comment by RSuter** [comment]
Another option would be to just provide template extension points so that you can fully customize the called methods...

**Comment by vojtechvit** [comment]
You're right that my suggestion expects the context classes to be generated. Ideally they should contain all info available about the endpoint being called + the API itself that was extracted from Open API specs - e.g. endpoint HTTP method, endpoint path, endpoint parameters, required OAuth scopes, related service security definitions etc. since all of this information may be useful when implementing smart generic interceptors.

Theoretically the context could be just a `JObject `of an expanded (no use of internal references to facilitate navigation) Open API specification - in such case the `OpenApiRequestContext.EndPoint` would point to the field of the endpoint being called while `OpenApiRequestContext.Api` would point to the root of the `JObject`. This would be a very generic solution that would require very few, if any, new classes. In any case, though, the (expanded) Open API specification should be stored in some string/resource file.

> Another option would be to just provide template extension points so that you can fully customize the called methods...

While it may be of value in some scenarios, I don't think this is a good solution for scenarios like adding Authentication header based on scopes that are defined in Open API specs on endpoint level - this can (should) be implemented as a generic interceptor so there should also be support for that ;-)

**Comment by RSuter** [comment]
We could use the SwaggerOperation class from NSwag.Core (i.e. the generated code dependes on the NSwag.Core package), generate just the JSON and deserialize on runtime to SwaggerOperation...

**Comment by yngndrw-sage** [comment]
I've also just run into the issue of needing asynchronous logic within the PrepareRequest method, in my case I need to retrieve an access token via the OAuth client credentials flow. (Which is cached, but access tokens can expire so I can't just retrieve it once)

**Comment by gzinger** [comment]
all I need is change the methods to async

**Comment by RicoSuter** [comment]
I think because PrepareRequest is a partial method it cannot be async because a partial method cannot have a return value (as it might not be implemented).

**Comment by johanndev** [comment]
Partial methods got some nice upgrades in C# 9 - they can now return values, see: https://blog.miguelbernard.com/c-9-extending-partial-methods

**Comment by Eugene-Vector** [comment]
Hello. Is this feature available now? I really need to access caller method context in `PrepareRequest()` and `ProcessResponse()` to store some extracted context-related data (request and response model types) in `request.Options` and access it later in the pipeline.

**Comment by John-Paul-R** [comment]
Since the "partial methods" blog post linked my @johanndev is now unreachable, a link to the [Extending partial methods](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/extending-partial-methods) proposal notes (which went in in C# 9)

Also, +1 on the feature. Would be valuable to me. Use case is adding a fixed (i.e. not a parameter) API version header on every request, depending on some per-request metadata.
